miniapps/nurbs/nurbs_ex25p.cpp

```diff
@@ -0,0 +1,753 @@
+   //     applying any necessary transformations such as: eliminating boundary
+   //     conditions, applying conforming constraints for non-conforming AMR,
+   //     static condensation, etc.
+
+   if (pa) { a.SetAssemblyLevel(AssemblyLevel::PARTIAL); }
+   a.Assemble(0);
+
+   OperatorPtr A;
+   Vector B, X;
+   a.FormLinearSystem(ess_tdof_list, x, b, A, X, B);
+
+   cout << "Size of linear system: " << A->Height() << endl;
+
+   ConstantCoefficient absomeg(pow2(omega) * epsilon);
+   RestrictedCoefficient restr_absomeg(absomeg,attr);
+
+   ParBilinearForm prec(fespace);
+   prec.AddDomainIntegrator(new CurlCurlIntegrator(restr_muinv));
+   prec.AddDomainIntegrator(new VectorFEMassIntegrator(restr_absomeg));
+
+   PMLDiagMatrixCoefficient pml_c1_abs(cdim,detJ_inv_JT_J_abs, pml);
+   ScalarVectorProductCoefficient c1_abs(muinv,pml_c1_abs);
+   VectorRestrictedCoefficient restr_c1_abs(c1_abs,attrPML);
+
+   PMLDiagMatrixCoefficient pml_c2_abs(dim, detJ_JT_J_inv_abs,pml);
+   ScalarVectorProductCoefficient c2_abs(absomeg,pml_c2_abs);
+   VectorRestrictedCoefficient restr_c2_abs(c2_abs,attrPML);
+
+   prec.AddDomainIntegrator(new CurlCurlIntegrator(restr_c1_abs));
+   prec.AddDomainIntegrator(new VectorFEMassIntegrator(restr_c2_abs));
+
+   prec.Assemble();
+
+   // 11. Define and apply a GMRES solver for AU=B with a block diagonal
+   //     preconditioner based on the Gauss-Seidel or Jacobi sparse smoother.
+   Array<int> offsets(3);
+   offsets[0] = 0;
+   offsets[1] = fespace->GetTrueVSize();
+   offsets[2] = fespace->GetTrueVSize();
+   offsets.PartialSum();
+
+   std::unique_ptr<Operator> pc_r;
+   std::unique_ptr<Operator> pc_i;
+   real_t s = (conv == ComplexOperator::HERMITIAN) ? -1_r : 1_r;
+
+   OperatorPtr PCOpAh;
+   prec.SetDiagonalPolicy(mfem::Operator::DIAG_ONE);
+   prec.FormSystemMatrix(ess_tdof_list, PCOpAh);
+
+
+   // Gauss-Seidel Smoother
+   pc_r.reset(new HypreBoomerAMG(*PCOpAh.As<HypreParMatrix>()));
+   pc_i.reset(new ScaledOperator(pc_r.get(), s));
+
+   BlockDiagonalPreconditioner BlockDP(offsets);
+   BlockDP.SetDiagonalBlock(0, pc_r.get());
+   BlockDP.SetDiagonalBlock(1, pc_i.get());
+
+   GMRESSolver gmres(MPI_COMM_WORLD);
+   gmres.SetPrintLevel(1);
+   gmres.SetKDim(200);
+   gmres.SetMaxIter(10000);
+   gmres.SetRelTol(1e-5);
+   gmres.SetAbsTol(0.0);
+   gmres.SetOperator(*A);
+   gmres.SetPreconditioner(BlockDP);
+   gmres.Mult(B, X);
+
+   // 12. Recover the solution as a finite element grid function.
+   a.RecoverFEMSolution(X, b, x);
+
+   ofstream mesh_ofs("nurbs_ex25p.mesh");
+   mesh_ofs.precision(8);
+
+   pmesh->Print(mesh_ofs);
+
+   ofstream sol_r_ofs("nurbs_ex25p-sol_r.gf");
+   ofstream sol_i_ofs("nurbs_ex25p-sol_i.gf");
+   sol_r_ofs.precision(8);
+   sol_i_ofs.precision(8);
+   x.real().Save(sol_r_ofs);
+   x.imag().Save(sol_i_ofs);
+
+   ParaViewDataCollection *pd = NULL;
+   pd = new ParaViewDataCollection("nurbs_ex25p", pmesh);
```

Comment:
Instead of `new/delete`, this can simply be an object: `ParaViewDataCollection pd(...`